src/coreclr/interpreter/compiler.cpp

```diff
@@ -4770,6 +4770,468 @@ constexpr uint8_t getLittleEndianByte(uint32_t value, int byteNum)
+};
+
+static OpcodePeepElement peepStLdLoc_S[] = {
+    { 0, CEE_STLOC_S },
+    { 2, CEE_LDLOC_S },
+    { 4, CEE_ILLEGAL } // End marker
+};
+
+static OpcodePeepElement peepStLdLoc[] = {
+    { 0, CEE_STLOC },
+    { 5, CEE_LDLOC },
+    { 10, CEE_ILLEGAL } // End marker
+};
+
+static OpcodePeepElement peepBoxUnboxOpcodes[] = {
+    { 0, CEE_BOX },
+    { 5, CEE_UNBOX_ANY },
+    { 10, CEE_ILLEGAL } // End marker
+};
+
+OpcodePeepElement peepTypeValueTypeOpcodes[] = {
+    { 0, CEE_LDTOKEN },
+    { 5, CEE_CALL },
+    { 10, CEE_CALL },
+    { 15, CEE_ILLEGAL } // End marker
+};
+
+class InterpILOpcodePeeps
+{
+public:
+    OpcodePeep peepTypeEqualityCheck = { peepTypeEqualityCheckOpcodes, &InterpCompiler::IsTypeEqualityCheckPeep, &InterpCompiler::ApplyTypeEqualityCheckPeep, "TypeEqualityCheck" };
+    OpcodePeep peepStoreLoad0 = { peepStLdLoc0, &InterpCompiler::IsStoreLoadPeep, &InterpCompiler::ApplyStoreLoadPeep, "StoreLoad0" };
+    OpcodePeep peepStoreLoad1 = { peepStLdLoc1, &InterpCompiler::IsStoreLoadPeep, &InterpCompiler::ApplyStoreLoadPeep, "StoreLoad1" };
+    OpcodePeep peepStoreLoad2 = { peepStLdLoc2, &InterpCompiler::IsStoreLoadPeep, &InterpCompiler::ApplyStoreLoadPeep, "StoreLoad2" };
+    OpcodePeep peepStoreLoad3 = { peepStLdLoc3, &InterpCompiler::IsStoreLoadPeep, &InterpCompiler::ApplyStoreLoadPeep, "StoreLoad3" };
+    OpcodePeep peepStoreLoadS = { peepStLdLoc_S, &InterpCompiler::IsStoreLoadPeep, &InterpCompiler::ApplyStoreLoadPeep, "StoreLoadS" };
+    OpcodePeep peepStoreLoad = { peepStLdLoc, &InterpCompiler::IsStoreLoadPeep, &InterpCompiler::ApplyStoreLoadPeep, "StoreLoad" };
+    OpcodePeep peepBoxUnbox = { peepBoxUnboxOpcodes, &InterpCompiler::IsBoxUnboxPeep, &InterpCompiler::ApplyBoxUnboxPeep, "BoxUnbox" };
+    OpcodePeep peepTypeValueType = { peepTypeValueTypeOpcodes, &InterpCompiler::IsTypeValueTypePeep, &InterpCompiler::ApplyTypeValueTypePeep, "TypeValueType" };
+
+public:
+    OpcodePeep* Peeps[10] = { 
+        &peepTypeEqualityCheck,
+        &peepStoreLoad,
+        &peepStoreLoad1,
+        &peepStoreLoad2,
+        &peepStoreLoad3,
+        &peepStoreLoadS,
+        &peepStoreLoad0,
+        &peepBoxUnbox,
+        &peepTypeValueType,
+        NULL };
+
+    bool FindAndApplyPeep(InterpCompiler* compiler)
+    {
+        const uint8_t* ip = compiler->m_ip;
+        
+        for (int i = 0; Peeps[i] != NULL; i++)
+        {
+            OpcodePeep *peep = Peeps[i];
+            bool skipToNextPeep = false;
+
+            // Check to see if peep applies to the current block just looking at the IL streams
+            // starting at the current ip.
+            for (int iPeepOpCode = 0; peep->pattern[iPeepOpCode].opcode != CEE_ILLEGAL; iPeepOpCode++)
+            {
+                int32_t insOffset = (int32_t)((ip + peep->pattern[iPeepOpCode].offsetIntoPeep) - compiler->m_pILCode);
+
+                if (ip + peep->pattern[iPeepOpCode].offsetIntoPeep >= compiler->m_pILCode + compiler->m_ILCodeSize)
```

Comment:
A nit - it would be nice to move the `const uint8_t *ipForOpcode = ip + peep->pattern[iPeepOpCode].offsetIntoPeep;` before this line and use it here too.